srf_generation/velocity_model_generation/generate_3d_velocity_model_perturbation.py

```diff
@@ -92,88 +114,124 @@ def main():
         for index, l_p in layer_params.items()
     ]
-    try:
-        with Pool(n_processes) as pool:
-            layer_info = sorted(pool.starmap(kwarg_map, complete_layer_parameters))
-    except AssertionError as e:
+    if n_processes == 1:
         layer_info = sorted([kwarg_map(*layer) for layer in complete_layer_parameters])
+    else:
+        try:
+            with Pool(n_processes) as pool:
+                layer_info = sorted(pool.starmap(kwarg_map, complete_layer_parameters))
+        except AssertionError:
+            print("The process failed at the layer generation, try with n_proc=1")
+            raise
     combine_layers(layer_info, common_params["nx"], common_params["ny"], out_file)
     for _, _, file in layer_info:
         remove(file)
 
 
+def set_verbose(verbose):
+    if verbose:
+        global DEBUG_OUTPUT_LOCATION
+        DEBUG_OUTPUT_LOCATION = None
+
+
 def generate_velocity_model_perturbation_file_from_model(
-    vm_params, perturbation_model, out_file, n_processes=1
+    vm_params, perturbation_model, out_file, n_processes=1, verbose=False
 ):
+    set_verbose(verbose)
     max_depth = vm_params["nz"] * vm_params["hh"]
     current_depth = 0
 
     i = 0
 
-    complete_layer_parameters = []
-    while current_depth < max_depth:
-        layer = perturbation_model.iloc[i]
-        layer_depth = layer["depth"]
-        if layer["depth"] + current_depth > max_depth:
-            layer_depth = max_depth - current_depth
-        nz = int(layer_depth / vm_params["hh"])
-        seed = randint(0, 2 ** 31 - 1)
-
-        complete_layer_parameters.append(
-            (
-                i,
-                vm_params["nx"],
-                vm_params["ny"],
-                nz,
-                vm_params["hh"],
-                layer["h_corr"],
-                layer["v_corr"],
-                layer["sigma"],
-                seed,
+    # Uses a temp dir in the current folder structure due to memory issues if using an actual temp directory
+    with TemporaryDirectory(dir=Path(out_file).parent) as temp_dir:
+        temp_dir_path = Path(temp_dir).resolve()
+        complete_layer_parameters = []
+        while current_depth < max_depth:
+            layer = perturbation_model.iloc[i]
+            layer_depth = layer["depth"]
+            if layer["depth"] + current_depth > max_depth:
+                layer_depth = max_depth - current_depth
+            nz = int(layer_depth / vm_params["hh"])
+            seed = randint(0, 2 ** 31 - 1)
+
+            complete_layer_parameters.append(
+                (
+                    i,
+                    vm_params["nx"],
+                    vm_params["ny"],
+                    nz,
+                    vm_params["hh"],
+                    layer["h_corr"],
+                    layer["v_corr"],
+                    layer["sigma"],
+                    seed,
+                    temp_dir_path,
```

Comment:
Question. When your code gets out of this "with TempoararyDirectory" scope, the temp directory gets deleted. When you do some action with complete_layer_parameters later with this "temp_dir_path", ie. create_perturbated_layer(), is it ok even if the temp directory doesn't actually exist?

Reply:
Yes. Temporary files are used by the combiner and then outputs a file in the output directory. A lot of the "functionality" is masked by the binaries and this script is really a wrapper for the binaries.

Reply:
That's interesting - does it mean temp directory name could have been just generated without being actually created?

Reply:
The function create_perturbated_layer requires the last parameter to be the path to a directory. Here it uses a temporary directory so that it automatically gets cleaned up after, but a regular directory can be used too. 
If you give the path to a directory that doesn't exist then Fractal3D (The layer creator C program) will crash as it can't create the output file.
The TempDirectroy is still in scope/exists when the pool is called with the complete_layer_parameters, so no issues there.
To the best of my knowledge create_perturbated_layer doesn't create folders either, if you do give it a directory that doesn't exist.

Reply:
@jpa198 True - github comparison view made me believe line 226 was outside the scope of TemporaryDirectory.  I'm happy